File: generator/src/test/resources/com/cromoteca/samples/endpoints/CustomTypesEndpoint.ts
import client from '../../../../connect-client.default.js';
import type MyLibraryType from '../../../example/library/MyLibraryType.js';
import type MyCustomType from './CustomTypesEndpoint/MyCustomType.js';
import type MyOtherType from './entities/MyOtherType.js';
import { type EndpointRequestInit } from '@hilla/frontend';

async function externalLibraryType(argument: MyLibraryType, init?: EndpointRequestInit): Promise<MyLibraryType> {
    return client.call('com.cromoteca.samples.endpoints.CustomTypesEndpoint', 'externalLibraryType', { argument }, init);
}

async function fromArrayOfCustomsToCustom(argument: Array<MyOtherType>, init?: EndpointRequestInit): Promise<MyOtherType> {
    return client.call('com.cromoteca.samples.endpoints.CustomTypesEndpoint', 'fromArrayOfCustomsToCustom', { argument }, init);
}

async function fromCustomToListOfCustoms(argument: MyCustomType, init?: EndpointRequestInit): Promise<Array<MyCustomType>> {
    return client.call('com.cromoteca.samples.endpoints.CustomTypesEndpoint', 'fromCustomToListOfCustoms', { argument }, init);
}

const CustomTypesEndpoint = {
    externalLibraryType,
    fromArrayOfCustomsToCustom,
    fromCustomToListOfCustoms,
};

export default CustomTypesEndpoint;

File: endpoints/src/main/java/com/cromoteca/samples/endpoints/CustomTypesEndpoint.java
package com.cromoteca.samples.endpoints;

import com.cromoteca.samples.endpoints.entities.MyOtherType;
import com.example.library.MyLibraryType;
import com.vaadin.flow.server.auth.AnonymousAllowed;
import com.vaadin.hilla.Endpoint;
import java.util.List;

@Endpoint
@AnonymousAllowed
public class CustomTypesEndpoint {
  public List<MyCustomType> fromCustomToListOfCustoms(MyCustomType argument) {
    return List.of(argument);
  }

  public MyOtherType fromArrayOfCustomsToCustom(MyOtherType[] argument) {
    return argument[0];
  }

  public MyLibraryType externalLibraryType(MyLibraryType argument) {
    return argument;
  }

  public static class MyCustomType {
    private String field;

    public String getField() {
      return field;
    }

    public void setField(String field) {
      this.field = field;
    }
  }
}
